### ISSUE TITLE
BUG: pdist cosine measure fails to return 0 distance for same vectors

### ISSUE THREAD

**Issue by joernhees** [scipy/scipy#5208]
The current cosine distance implementation fails to return a distance of 0 when asked to compare a vector with itself.

The following happens in scipy 0.16.0 on Mac OS X 10.9.5, homebrew python 2.7.10, pip installed numpy and scipy:

```python
In [1]: import numpy as np

In [2]: import scipy as sc

In [3]: from scipy.spatial.distance import pdist, squareform

In [4]: np.random.seed(1337)

In [5]: a = np.random.rand(500)

In [6]: a = np.vstack((a, a))  # same vector twice

In [7]: for d in range(2, 501):
   ...:     print d, pdist(a[:, :d]), pdist(a[:, :d], 'cosine'), pdist(a[:, :d], 'old_cosine')
   ...:
2 [ 0.] [  1.11022302e-16] [  1.11022302e-16]
3 [ 0.] [ 0.] [ 0.]
4 [ 0.] [ 0.] [ 0.]
5 [ 0.] [  1.11022302e-16] [  1.11022302e-16]
6 [ 0.] [ -2.22044605e-16] [ -2.22044605e-16]
7 [ 0.] [ -2.22044605e-16] [ -2.22044605e-16]
8 [ 0.] [ 0.] [ -2.22044605e-16]
9 [ 0.] [ -2.22044605e-16] [ -2.22044605e-16]
10 [ 0.] [  2.22044605e-16] [  2.22044605e-16]
...
```

As you can see the default euclidean distance is computed correctly, but both cosine distances behave erratic (and different). This is probably a floating point precision problem, but it seems as if there is special / better handling for identical vectors in the euclidean case.

Especially the returned negative distances are quite annoying if used for example via `Z = linkage(..., metric='cosine')`. The returned linkage matrix `Z` will most likely contain some negative values if the same feature vectors are encountered several times (which is kind of common and a low hanging fruit in every clustering). Methods like `dendrogram(Z)` will check `Z` and throw an error as there are negative distances in it.

It would be great if `pdist(..., 'cosine')` would check for identical values before running its more complicated calculations and correctly return 0 in those cases. If for whatever reason this seems unfeasible, it would be great if the results would at least be clipped to `[0,1]` (as suggested [here](http://stackoverflow.com/questions/2590117/scipy-negative-distance-what/10901549#10901549)).

**Comment by larsmans** [comment]
This happens because, in this case, two different methods of computing the squared norm are used. The implementation of cosine distance is

```py
1.0 - np.dot(u, v) / (norm(u) * norm(v))
```

and `dot` uses BLAS's `ddot`, but `norm` uses `dnrm2`. The latter typically uses a more stable algorithm.

**Comment by larsmans** [comment]
My (deleted) comment about the implementation was not entirely correct. The implementation of `pdist`'s cosine distance is [here](/scipy/scipy/blob/master/scipy/spatial/src/distance_impl.h#L411). The problem is still a discrepancy between two ways of computing the squared norm x⋅x, though.

It's easy enough to at least clip the values; negative distances make no sense.

**Comment by WarrenWeckesser** [comment]
Just a note to point out that, while in theory `cosine(x, -x)` should be exactly 2, normal floating point imprecision can result in values less than or greater than 2:

```
In [109]: x = np.arange(10.)

In [110]: cosine(x, -x)
Out[110]: 1.9999999999999998

In [111]: np.random.seed(1)

In [112]: x = np.arange(10000.) + np.random.randn(10000)

In [113]: cosine(x, -x)
Out[113]: 2.0000000000000009
```
Of course, this is much less of a nuisance than the negative values that `cosine(x, x)` can return, but it might be worth considering clipping the result of the cosine distance to the interval `[0, 2]` (not `[0, 1]` as suggested earlier).

**Comment by larsmans** [comment]
Good point, I hadn't actually considered negative values in #5225. But it's hard to devise a good test for this: on my box, the last `cosine(x, -x)` actually returns `1.9999999999999998`.

**Comment by larsmans** [comment]
Updated #5225.

**Comment by joernhees** [comment]
Oups, sorry, in my case all the dimensions are >= 0 (clustering word counts), which is why the result should be in [0, 1], but in general you're obviously right, they can be in [0, 2].

The clipping fix looks fine for my purposes. I can live with it if two identical vectors have a distance of `0 (+ epsilon)`, exactly perpendicular vectors have a distance of `1 (+- epsilon)` and two inverted vectors have a distance of `2 (- epsilon)`.

I guess it wouldn't be hard to additionally check for identical or inverted vectors, and in those cases return the exact distances (`0` or `2`), but i guess it's debatable if this belongs into scipy:

Pros:
- exact results for identical or inverted vectors
- probably faster than `cosine = dot_product(u, v, n) / (norms[i] * norms[j]);` in those cases

Cons:
- inconsistent: other cos-distances aren't exact either (most notably perpendicular ones around dist `1`)
- additional overhead for general case

That being said, @larsmans thanks for the clipping and feel free to close this...

**Comment by larsmans** [comment]
If you want exact results for equal and opposite vectors, then I guess the recipe for client code is to normalize the vectors in-place and compare them by the ordinary dot product.

**Comment by larsmans** [comment]
I merged #5225. Closing this one.
